### ISSUE TITLE
Non-synthetic Key.alt and Key.cmd

### ISSUE THREAD

**Issue by kenjitoyama** [moses-palmer/pynput#4]
When I press Key.alt + Key.tab I'm expecting my OS (Ubuntu) to switch windows, but nothing happens.

When I start xev and issue Key.alt I see:

KeyPress event, serial 37, synthetic YES, window 0x5400001,
    root 0x498, subw 0x0, time 0, (0,0), root:(0,0),
    state 0x8, keycode 64 (keysym 0xffe9, Alt_L), same_screen NO,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

while if I issue a real Alt on my qwerty keyboard I see:

KeyPress event, serial 37, synthetic NO, window 0x5400001,
    root 0x498, subw 0x0, time 26110842, (20,-9), root:(85,43),
    state 0x0, keycode 64 (keysym 0xffe9, Alt_L), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

Most things are similar, but I noticed that "synthetic" is different. Is this the reason why I can't use pynput to control the OS? The behavior is the same with Key.cmd (Super/Win). FYI if I use PyUserInput it generates "synthetic NO" events in xev.

**Comment by moses-palmer** [comment]
I have pushed a fix to the *master* branch.

Instead of always using `XSendEvent`, which is necessary to handle multiple layouts, *pynput* no uses *XFAKE* for control keys.

**Comment by kenjitoyama** [comment]
Thanks for the quick fix, Moses. Now Key.ctrl, Key.shift, Key.alt and Key.cmd send non-synthetic keys. I can use Key.cmd for example to trigger the application menu.

However, Key.tab still sends a synthetic key, which still doesn't trigger application switching:

KeyRelease event, serial 38, synthetic YES, window 0x4e00001,
    root 0x498, subw 0x0, time 0, (0,0), root:(0,0),
    state 0x0, keycode 23 (keysym 0xff09, Tab), same_screen NO,
    XLookupString gives 1 bytes: (09) " "
    XFilterEvent returns: False


Side question: What happens if all keys, including normal letter keys, are sent as non-synthetic?

**Comment by moses-palmer** [comment]
I quickly checked the documentation for (XSendEvent)[https://tronche.com/gui/x/xlib/event-handling/XSendEvent.html] and [XKeyEvent](https://tronche.com/gui/x/xlib/events/keyboard-pointer/keyboard-pointer.html#XKeyEvent), and it does not seem to support manipulating the *synthetic* flag. *PyUserInput* uses *XFAKE*, which is a server extension, for sending fake events; I presume this is why it can set the flag.

Normal letters cannot be sent through *XFAKE* if a general solution compatible with the *pynput API guarantees* is sought; *XFAKE* does not support sending keys not mapped in the layout, and maintains its own modifier mapping.

I have pushed a new set of commits; now all keys with a known virtual key code are sent through *XFAKE*, thus allowing control keys and modifiers to be non-synthetic. I extended my local tests with alt+tab, and that too works now.

**Comment by kenjitoyama** [comment]
Hey, you're very quick! It works beautifully now. =) Thanks very much!

**Comment by kenjitoyama** [comment]
Hmm, something has changed though. Example:

Python 2.7.12 (default, Jul  1 2016, 15:12:24) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from pynput.keyboard import Key, Controller
>>> 
>>> keyboard = Controller()
>>> keyboard.press('a')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build/bdist.linux-x86_64/egg/pynput/keyboard/_base.py", line 371, in press
  File "build/bdist.linux-x86_64/egg/pynput/keyboard/_xorg.py", line 204, in _handle
  File "/usr/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "build/bdist.linux-x86_64/egg/pynput/_util/xorg.py", line 76, in display_manager
pynput._util.xorg.X11Error: [(BadValue(), None)]

I installed pynput just now from 024dfd703d9da7a404ffcdac461b8cea444fbe25

**Comment by moses-palmer** [comment]
I had a feeling this would be one of those ;-)

I made the mistake of not running the test suite after the modifications; it was so small, right?

I have pushed yet another set of commits that changes the default value of [pynput.KeyCode.vk](https://github.com/moses-palmer/pynput/blob/master/lib/pynput/keyboard/_base.py#L42) to `None`, which is a saner value, and what the previous patch expected.

Thank you for your input and testing!
